source/extensions/transport_sockets/tls/ocsp/asn1_utility.h

```diff
@@ -0,0 +1,208 @@
+#pragma once
+
+#include <iomanip>
+#include <sstream>
+#include <vector>
+
+#include "envoy/common/exception.h"
+#include "envoy/common/time.h"
+
+#include "common/common/assert.h"
+
+#include "absl/types/optional.h"
+#include "openssl/bn.h"
+#include "openssl/bytestring.h"
+#include "openssl/ssl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace Tls {
+namespace Ocsp {
+
+/**
+ * Construct a |T| from the data contained in the CBS&. Functions
+ * of this type must advance the input CBS& over the element.
+ */
+template <typename T> using Asn1ParsingFunc = std::function<T(CBS&)>;
+
+/**
+ * Utility functions for parsing DER-encoded ASN.1 objects.
+ * This relies heavily on the 'openssl/bytestring' API which
+ * is BoringSSL's recommended interface for parsing DER-encoded
+ * ASN.1 data when there is not an existing wrapper.
+ * This is not a complete library for ASN.1 parsing and primarily
+ * serves as abstractions for the OCSP module, but can be
+ * extended and moved into a general utility to support parsing of
+ * additional ASN.1 objects.
+ *
+ * Each function adheres to the invariant that given a reference
+ * to a crypto bytestring (CBS&), it will parse the specified
+ * ASN.1 element and advance `cbs` over it.
+ *
+ * An exception is thrown if the bytestring is malformed or does
```

Comment:
I've seen some other utilities in envoy using a FooResult type as a struct with a result and an error. Maybe that would be a good option here so we can get error strings?

Reply:
If you want the error string as well, use a `variant`. Here's an example: `using VariantHeader = absl::variant<absl::string_view, InlineHeaderVector>;`.

But in many/most of these cases, there's only 1 way it can fail, so it may not be needed.

Reply:
Ah ok, I think I misunderstood the reach of this change. Are you suggesting to return optionals from just `Asn1Utility` methods and still throw exceptions in the OCSP-specific parsing methods (which are currently only used at config time)?

Reply:
Yes, anything that you know will only be called at config time, an exception is fine.

Reply:
One thing I just realized, how would the `absl::optional` approach work with return values that are already optionals? How could we differentiate the error case from the missing case?

Reply:
Oh, so you really need a tri-state: not present, present, or error? Probably a variant would make sense then, maybe containing an optional and an error? Or a variant of 3 possible types, but one of them is essentially `void`. I'm not sure what the best of representing that is.

Reply:
I think a variant of two types, the return type or error, would work. In some cases the return type would just be an optional. This is basically Rust's `Result` type and I think takes care of the ambiguity problem.